File: libs/hikyuu_python/trade_sys/_Signal.cpp
/*
 * _Signal.cpp
 *
 *  Created on: 2013-3-18
 *      Author: fasiondog
 */

#include <boost/python.hpp>
#include <hikyuu/trade_sys/signal/SignalBase.h>
#include <hikyuu/trade_sys/signal/build_in.h>
#include "../pickle_support.h"

using namespace boost::python;
using namespace hku;

class SignalWrap : public SignalBase, public wrapper<SignalBase> {
public:
    SignalWrap(const string& name): SignalBase(name) {}

    void _reset() {
        this->get_override("_reset")();
    }

    SignalPtr _clone() {
        return this->get_override("_clone")();
    }

    void _calculate() {
        this->get_override("_calculate");
    }
};

BOOST_PYTHON_FUNCTION_OVERLOADS(AMA_SG_overload, AMA_SG, 0, 6);

void export_Signal() {
    class_<SignalWrap, boost::noncopyable>("SignalBase", init<const string&>())
            .def(self_ns::str(self))
            .add_property("name",
                    make_function(&SignalBase::name,
                            return_value_policy<copy_const_reference>()))
            .add_property("params",
                    make_function(&SignalBase::getParameter,
                            return_internal_reference<>()))
            .def("setTO", &SignalBase::setTO)
            .def("shouldBuy", &SignalBase::shouldBuy)
            .def("shouldSell", &SignalBase::shouldSell)
            .def("getBuySignal", &SignalBase::getBuySignal)
            .def("getSellSignal", &SignalBase::getSellSignal)
            .def("_addBuySignal", &SignalBase::_addBuySignal)
            .def("_addSellSignal", &SignalBase::_addSellSignal)
            .def("reset", &SignalBase::reset)
            .def("clone", &SignalBase::clone)
            .def("_calculate", pure_virtual(&SignalBase::_calculate))
            .def("_reset", pure_virtual(&SignalBase::_reset))
            .def("_clone", pure_virtual(&SignalBase::_clone))
#if HKU_PYTHON_SUPPORT_PICKLE
            .def_pickle(name_init_pickle_suite<SignalBase>())
#endif
            ;

    register_ptr_to_python<SignalPtr>();


    def("AMA_SG", AMA_SG, AMA_SG_overload());

}

File: tools/hikyuu/test/Signal.py
#!/usr/bin/python
# -*- coding: utf8 -*-
# gb18030

#===============================================================================
# 作者：fasiondog
# 历史：1）20130321, Added by fasiondog
#===============================================================================

import unittest

from test_init import *
from hikyuu.trade_sys.signal import *

class SignalPython(SignalBase):
    def __init__(self):
        super(SignalPython, self).__init__("SignalPython")
        self._x = 0
    
    def _reset(self):
        self._x = 0
            
    def _clone(self):
        p = SignalPython()
        p._x = self._x
        return p

class SignalTest(unittest.TestCase):
    def test_SignalBase(self):
        p = SignalPython()
        self.assertEqual(p.name, "SignalPython")
        
        self.assertEqual(p.shouldBuy(Datetime(200101010000)), False)
        p._addBuySignal(Datetime(200101010000))
        self.assertEqual(p.shouldBuy(Datetime(200101010000)), True)
        
        self.assertEqual(p.shouldSell(Datetime(200101010000)), False)
        p._addSellSignal(Datetime(200101010000))
        self.assertEqual(p.shouldSell(Datetime(200101010000)), True)
        
        self.assertEqual(p._x, 0)
        p._x = 10
        self.assertEqual(p._x, 10)
        p.reset()
        self.assertEqual(p._x, 0)
        
        p._x = 10
        p_clone = p.clone()
        self.assertEqual(p_clone._x, 10)
        p.reset()
        self.assertEqual(p._x, 0)
        self.assertEqual(p_clone._x, 10)

                 
def suite():
    return unittest.TestLoader().loadTestsFromTestCase(SignalTest)